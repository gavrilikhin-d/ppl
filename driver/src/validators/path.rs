use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
