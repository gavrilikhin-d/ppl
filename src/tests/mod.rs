use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
