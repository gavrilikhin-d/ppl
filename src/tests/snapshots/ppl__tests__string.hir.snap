---
source: src/tests/mod.rs
expression: hir
---
let $tmp@8: String = `<:String> + <:String>`(`<:String> + <:String>`("Hello", " "), "World!")
`println <:Reference<String>>`((&$tmp@8:Reference<String>))
let $tmp@42: String = `String from <:None>`(none)
`println <:Reference<String>>`((&$tmp@42:Reference<String>))
let $tmp@69: String = `String from <:Bool>`(false)
`println <:Reference<String>>`((&$tmp@69:Reference<String>))
let $tmp@97: String = `String from <:Bool>`(true)
`println <:Reference<String>>`((&$tmp@97:Reference<String>))
let $tmp@124: String = `String from <:Integer>`(0)
`println <:Reference<String>>`((&$tmp@124:Reference<String>))
let $tmp@148: String = `String from <:Rational>`(0.0)
`println <:Reference<String>>`((&$tmp@148:Reference<String>))
let $tmp@174: String = `String from <:Type<Integer>>`((copy Type<Integer>:Type<Integer>))
`println <:Reference<String>>`((&$tmp@174:Reference<String>))
let $tmp@204: String = `String from <:Type<Array<Integer>>>`((copy Type<Array<Integer>>:Type<Array<Integer>>))
`println <:Reference<String>>`((&$tmp@204:Reference<String>))
`destroy <:ReferenceMut<String>>`(($tmp@8:String))
`destroy <:ReferenceMut<String>>`(($tmp@42:String))
`destroy <:ReferenceMut<String>>`(($tmp@69:String))
`destroy <:ReferenceMut<String>>`(($tmp@97:String))
`destroy <:ReferenceMut<String>>`(($tmp@124:String))
`destroy <:ReferenceMut<String>>`(($tmp@148:String))
`destroy <:ReferenceMut<String>>`(($tmp@174:String))
`destroy <:ReferenceMut<String>>`(($tmp@204:String))

==MONOMORPHIZED==


fn<Integer> String from <ty: Type<Integer>> -> String:
	let $tmp@4629: String = `clone <:Reference<String>>`((ty:Type<Integer>).name)
	return ($tmp@4629:String)


fn<Array<Integer>> String from <ty: Type<Array<Integer>>> -> String:
	let $tmp@4629: String = `clone <:Reference<String>>`((ty:Type<Array<Integer>>).name)
	return ($tmp@4629:String)
